### PR TITLE
Make Ctrl work with non-terminal apps too

### PR DESCRIPTION
**Pull request by mirfatif**

### CODE REVIEW

**Comment by menny** [comment]
this is great.
A couple of things:
1) can you add some comments and remove the extra logging?
2) can you add unit-tests for this?

Thank you

**Comment by mirfatif** [comment]
@menny please check.

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java [hunk 1]
```diff
@@ -667,15 +685,29 @@ private void onNonFunctionKey(
             default:
                 if (isWordSeparator(primaryCode)) {
                     handleSeparator(primaryCode);
-                } else {
-                    if (mControlKeyState.isActive() && primaryCode >= 32 && primaryCode < 127) {
+                } else if (mControlKeyState.isActive()) {
+                    int keyCode = getKeyCode(primaryCode);
+                    if (BuildConfig.DEBUG)
```
Comment (menny) [inline ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java:690]: No need for these logs. Please remove as many loggings as makes sense

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java [hunk 2]
```diff
@@ -667,15 +685,29 @@ private void onNonFunctionKey(
...
+                        Logger.d(TAG, "CONTROL state: Char code is " + primaryCode);
+                    if (Build.VERSION.SDK_INT >= 11 && keyCode != 0) {
+                        // TextView (and hence its subclasses) can handle ^A, ^Z, ^X, ^C and ^V
+                        // https://android.googlesource.com/platform/frameworks/base/+/refs/tags/android-10.0.0_r1/core/java/android/widget/TextView.java#11136
+                        int down = KeyEvent.ACTION_DOWN;
+                        int up = KeyEvent.ACTION_UP;
+                        int ctrlKey = KeyEvent.KEYCODE_CTRL_LEFT;
```
Comment (menny) [inline ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java:697]: Please use these constants directly

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java [hunk 3]
```diff
@@ -684,6 +716,8 @@ private void onNonFunctionKey(
                     } else {
                         handleCharacter(primaryCode, key, multiTapIndex, nearByKeyCodes);
                     }
+                } else {
+                    handleCharacter(primaryCode, key, multiTapIndex, nearByKeyCodes);
```
Comment (menny) [inline ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java:720]: Why do we need this call now? handleCharacter was not called before

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java [hunk 4]
```diff
@@ -1141,7 +1173,6 @@ public void onPress(int primaryCode) {
         if (primaryCode == KeyCodes.CTRL) {
             mControlKeyState.onPress();
             handleControl();
-            sendKeyDown(ic, 113); // KeyEvent.KEYCODE_CTRL_LEFT (API 11 and up)
```
Comment (menny) [inline ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java:1144]: So, we're holding up on this event until the user is pressing another key? What happens in the case where an app does want to react to CTRL press?

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java (hunk 1 above)
Comment (mirfatif) [reply ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java]: Done

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java (hunk 2 above)
Comment (mirfatif) [reply ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java]: Done

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java (hunk 3 above)
Comment (mirfatif) [reply ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java:714]: First call handles character when CTRL is pressed but character is out of range e.g. less than 32.
Second calls handles character when CTRL isn't pressed.

File: ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java (hunk 4 above)
Comment (mirfatif) [reply ime/app/src/main/java/com/anysoftkeyboard/AnySoftKeyboard.java]: I assumed that CTRL is always used with another key. That's how CTRL is used most of the times on Windows, Linux too. Reverted the changes back.

**Comment by mirfatif** [comment]
> Also, do you think we need any additional unit-tests?

As I said, a mechanical engineer is not the right person to answer this. 🙂

Also you may make any changes you want (in this PR or others of mine) before merging. I've been using ASK for quite some while and made changes over time according to my needs. I pushed this thinking if my little contribution might add some value to this great project. You may drop the unwanted parts or change them as needed. I won't mind. I'm always available for assistance.

**Pull request merged**